### DISCUSSION TITLE
Clarifications on User Profile Data

### DISCUSSION THREAD

**Discussion by Shakthieshwari** [Sunbird-Lern/Community#12] [Query]
Hello Team,

I have few clarifications on the User Profile Data , From the Read User API Response :- Please check the below response attached

1. How do i get to know if a user is SSO User ?

{"webPages":"","maskedPhone":"******1501","tcStatus":"","loginId":"","subject":"","channel":"dikshacustodian","profileUserTypes":[{"type":"administrator","subType":"at"},{"type":"administrator","subType":"principal"}],"language":"","updatedDate":"2022-05-19 05:49:31:287+0000","password":"","managedBy":"","flagsValue":0,"id":"1b646900-c1f3-4a91-992b-b5a8d0787760","recoveryEmail":"","identifier":"1b646900-c1f3-4a91-992b-b5a8d0787760","thumbnail":"","profileVisibility":"","updatedBy":"1b646900-c1f3-4a91-992b-b5a8d0787760","accesscode":"","externalIds":[],"registryId":"","roleList":[{"name":"Book Creator","id":"BOOK_CREATOR"},{"name":"Membership Management","id":"MEMBERSHIP_MANAGEMENT"},{"name":"Flag Reviewer","id":"FLAG_REVIEWER"},{"name":"Report Viewer","id":"REPORT_VIEWER"},{"name":"Program Manager","id":"PROGRAM_MANAGER"},{"name":"Program Designer","id":"PROGRAM_DESIGNER"},{"name":"System Administration","id":"SYSTEM_ADMINISTRATION"},{"name":"Content Curation","id":"CONTENT_CURATION"},{"name":"Book Reviewer","id":"BOOK_REVIEWER"},{"name":"Content Creator","id":"CONTENT_CREATOR"},{"name":"Org Management","id":"ORG_MANAGEMENT"},{"name":"Course Admin","id":"COURSE_ADMIN"},{"name":"Org Moderator","id":"ORG_MODERATOR"},{"name":"Public","id":"PUBLIC"},{"name":"Admin","id":"ADMIN"},{"name":"Course Mentor","id":"COURSE_MENTOR"},{"name":"Content Reviewer","id":"CONTENT_REVIEWER"},{"name":"Report Admin","id":"REPORT_ADMIN"},{"name":"Org Admin","id":"ORG_ADMIN"}],"rootOrgId":"0126684405014528002","prevUsedEmail":"","firstName":"TAJAMMUL HUSSAIN","isMinor":false,"tncAcceptedOn":1637124389439,"allTncAccepted":{},"profileDetails":"","phone":"******1501","dob":"1972-12-31","grade":"","currentLoginTime":"","userType":"administrator","status":1,"lastName":"","tncLatestVersion":"v12","gender":"","prevUsedPhone":"","stateValidated":false,"encEmail":"Jemoljs1sbmFAW2wVlcFxv7UUHIwGKDzCLtesFFwN41ymVeJ14AeGkoNU5CN252s5uciH9vudE3i\na2cieP6fKxQDoyKWQuqLH0ho6zLgPABYc0+HgGo3C7u37QtrcrGz/tJR3ZPAAFSTyJ9yFc8WHcvy\nG0t0HiJjD+xEIZgJ6RQ=","isDeleted":false,"organisations":[{"organisationId":"0126684405014528002","approvedBy":"","channel":"dikshacustodian","updatedDate":"","approvaldate":"","isSystemUpload":false,"isDeleted":false,"id":"0131983266990325762811","isApproved":"","orgjoindate":"2021-01-20 06:11:44:047+0000","isSelfDeclaration":true,"updatedBy":"","orgName":"DIKSHA Custodian Org","addedByName":"","addedBy":"","associationType":2,"locationIds":["4ba4bdfb-d478-45f1-ba30-f8eb0aa03914"],"orgLocation":[{"id":"4ba4bdfb-d478-45f1-ba30-f8eb0aa03914","type":"state"}],"externalId":"dikshacustodian","userId":"1b646900-c1f3-4a91-992b-b5a8d0787760","isSchool":false,"hashTagId":"0126684405014528002","isSSO":false,"isRejected":"","locations":[{"code":"09","name":"Uttar Pradesh","id":"4ba4bdfb-d478-45f1-ba30-f8eb0aa03914","type":"state","parentId":""}],"position":"","orgLeftDate":""},{"organisationId":"01323882407510835244483","approvedBy":"","channel":"br","updatedDate":"2022-05-19 05:49:31:297+0000","approvaldate":"","isSystemUpload":false,"isDeleted":false,"id":"01323882407510835244483","isApproved":false,"orgjoindate":"2021-05-30 14:26:24:980+0000","isSelfDeclaration":true,"updatedBy":"1b646900-c1f3-4a91-992b-b5a8d0787760","orgName":"GMS SIMRA","addedByName":"","addedBy":"1b646900-c1f3-4a91-992b-b5a8d0787760","associationType":2,"locationIds":["645c54a4-6b75-4b77-8978-31aa489b638c","ef15abb9-e846-4da5-a05e-3dd4311d01a5","5784b8bf-127b-4a0f-a07d-8573c2bd2a9a","0653ffc0-cd43-4e9e-9d91-c89f49eac7b2"],"orgLocation":[{"id":"645c54a4-6b75-4b77-8978-31aa489b638c","type":"state"},{"id":"ef15abb9-e846-4da5-a05e-3dd4311d01a5","type":"block"},{"id":"5784b8bf-127b-4a0f-a07d-8573c2bd2a9a","type":"cluster"},{"id":"0653ffc0-cd43-4e9e-9d91-c89f49eac7b2","type":"district"}],"externalId":"10181501501","userId":"1b646900-c1f3-4a91-992b-b5a8d0787760","isSchool":true,"hashTagId":"01323882407510835244483","isSSO":false,"isRejected":false,"locations":[{"code":"10","name":"Bihar","id":"645c54a4-6b75-4b77-8978-31aa489b638c","type":"state","parentId":""},{"code":"101815","name":"VAISHALI","id":"ef15abb9-e846-4da5-a05e-3dd4311d01a5","type":"block","parentId":"0653ffc0-cd43-4e9e-9d91-c89f49eac7b2"},{"code":"1018150012","name":"GMS CHAKALHDAD","id":"5784b8bf-127b-4a0f-a07d-8573c2bd2a9a","type":"cluster","parentId":"ef15abb9-e846-4da5-a05e-3dd4311d01a5"},{"code":"1018","name":"Vaishali","id":"0653ffc0-cd43-4e9e-9d91-c89f49eac7b2","type":"district","parentId":"645c54a4-6b75-4b77-8978-31aa489b638c"}],"position":"","orgLeftDate":""}],"provider":"","countryCode":"+91","tncLatestVersionUrl":"https://ntpproductionall.blob.core.windows.net/termsandcond/terms-and-conditions-v12.html","maskedEmail":"ta***************@gmail.com","tempPassword":"","email":"ta***************@gmail.com","rootOrg":{"dateTime":"","preferredLanguage":"","keys":{},"organisationSubType":"","channel":"dikshacustodian","approvedBy":"","description":"DIKSHA Custodian Organization","updatedDate":"2021-09-29 13:09:52:248+0000","organisationType":5,"addressId":"","orgType":"","isTenant":true,"provider":"","locationId":"","orgCode":"","theme":"","id":"0126684405014528002","communityId":"","isApproved":"","isBoard":true,"email":"","slug":"dikshacustodian","isSSOEnabled":"","thumbnail":"","orgName":"DIKSHA Custodian Org","updatedBy":"50e297da-0c73-4994-8381-0bfdcd8d1b44","locationIds":[],"externalId":"dikshacustodian","orgLocation":[{"id":"4ba4bdfb-d478-45f1-ba30-f8eb0aa03914","type":"state"}],"isRootOrg":true,"rootOrgId":"0126684405014528002","imgUrl":"","approvedDate":"","homeUrl":"","orgTypeId":"","isDefault":"","createdDate":"2019-01-02 14:44:48:951+0000","createdBy":"","parentOrgId":"","hashTagId":"0126684405014528002","noOfMembers":"","status":1},"phoneVerified":true,"profileSummary":"","tcUpdatedDate":"","userLocations":[{"code":"1018","name":"Vaishali","id":"0653ffc0-cd43-4e9e-9d91-c89f49eac7b2","type":"district","parentId":"645c54a4-6b75-4b77-8978-31aa489b638c"},{"code":"1018150012","name":"GMS CHAKALHDAD","id":"5784b8bf-127b-4a0f-a07d-8573c2bd2a9a","type":"cluster","parentId":"ef15abb9-e846-4da5-a05e-3dd4311d01a5"},{"code":"10","name":"Bihar","id":"645c54a4-6b75-4b77-8978-31aa489b638c","type":"state","parentId":""},{"code":"101815","name":"VAISHALI","id":"ef15abb9-e846-4da5-a05e-3dd4311d01a5","type":"block","parentId":"0653ffc0-cd43-4e9e-9d91-c89f49eac7b2"},{"code":"10181501501","name":"GMS SIMRA","id":"01323882407510835244483","type":"school","parentId":""}],"recoveryPhone":"******7169","avatar":"","userName":"tajammulhusain_n8f7","userId":"1b646900-c1f3-4a91-992b-b5a8d0787760","userSubType":"at","declarations":[{"persona":"default","errorType":"","orgId":"012593397993308160261","status":"SUBMITTED","info":{"declared-email":"tajammulhusain531@gmail.com","declared-ext-id":"1018150150101","declared-phone":"9934831501","declared-school-name":"GMS SIMRA","declared-school-udise-code":"10181501501"}}],"promptTnC":false,"emailVerified":true,"lastLoginTime":0,"createdDate":"2021-01-20 06:11:44:036+0000","framework":{"board":["State (Bihar)"],"id":["br_k-12"]},"createdBy":"","profileUserType":{"subType":"at","type":"administrator"},"encPhone":"WihNgwH62iriGQu4e+cSP/DNcfAG8x8viCoZQP34DdDYahWo8rFOEIIj3rlj7KC8UZF1q7lrHL2m\nMvA1A/QSzoFN5xNc5YbA1Ic0ftRa8K8q0iAvC3CBkNhWSnbNihV7T6a+wzaAmCWueMEdPmZuRg==","location":"","tncAcceptedVersion":"v12"}

**And also How do we get :-**
1. No of Unique Users Logged in with HT and other official Role
2. No of Users with HT and Other Official Role Today

@reshmi-nair Can you please help me out with these queries ?

Cc- @aks30, @kiranharidas187, @vijiurs

thanks

**Comment by Shakthieshwari** [comment]
Team, Any update on this please ?

**Reply by reshmi-nair** [reply]
@Shakthieshwari We have to check the RootOrgId to see whether user is SSO user or not.
Q1. No of Unique Users Logged in with HT and other official Role - pls use user search api with "profileUserTypes.subType": "HT" in filter.
Q2. Did not understand this query. How it differs from Q1?

**Reply by Shakthieshwari** [reply]
@reshmi-nair Q1. what do i check in **RootOrgId** Json Object ?
**"isTenant":True** should i check for this

**Reply by reshmi-nair** [reply]
@Shakthieshwari Get the custodianOrgId from systemSettings and if it is not matching with rootOrgId in User object or id in RootOrgId Object, then it is a state user.

**Reply by pramodkvarma** [reply]
When it comes to Sunbird, what's the meaning of "state" user? @reshmi-nair 

@vrayulu @rhwarrier

**Reply by Shakthieshwari** [reply]
@reshmi-nair Is it possible to get User PII(UserName,Mobile No,Email Id..) from any other place ? because API might fail if we query for more users repeatedly.

**Reply by pramodkvarma** [reply]
I looked at JIRA story (your post against SB-Obsrv) and it doesn't explain the actual need for program manager to see everyone's phone and email. @Shakthieshwari can you please help me understand? What does she/he do seeing a list of people's email/phone? I am curious.

**Reply by pramodkvarma** [reply]
> When it comes to Sunbird, what's the meaning of "state" user? 

@reshmi-nair can you answer my above question please?

@vrayulu @rhwarrier

**Reply by Shakthieshwari** [reply]
@Prateek-slokam @aishwaryashikshalokam Can you please answer @pramodkvarma Query?

**Reply by reshmi-nair** [reply]
@pramodkvarma In the context of SB, a 'state user' would be a non-custodian user.

**Reply by rhwarrier** [reply]
> > When it comes to Sunbird, what's the meaning of "state" user?
> 
> @reshmi-nair can you answer my above question please?
> 
> @vrayulu @rhwarrier

@pramodkvarma  The differentiation between a 'state' user and a 'self signed-up' user was created for the context of Diksha. However, if we were to translate that to Sunbird, it could be viewed as Custodian Org users (self signed up) vs. Other Org users.

**Reply by pramodkvarma** [reply]
@reshmi-nair @rhwarrier then can we make sure Sunbird vocabulary is corrected appropriately instead of speaking DIKSHA language? This is critical differentiation. Agree?

**Reply by reshmi-nair** [reply]
Yes @pramodkvarma. We will keep this in mind during discussions.

**Reply by Shakthieshwari** [reply]
> @reshmi-nair Is it possible to get User PII(UserName,Mobile No,Email Id..) from any other place ? because API might fail if we query for more users repeatedly.

@reshmi-nair can you please answer to my query? Is it possible to get UserName from any other place ? because API might fail if we query for more users repeatedly.

**Reply by reshmi-nair** [reply]
@Shakthieshwari Can you please share the usecase?

**Reply by Shakthieshwari** [reply]
@reshmi-nair We need to get the UserName of the users but not from the API, because it will fail if we query fro more users repeatedly at the same time. so if their is anyother way we can get this out either from some DB (or) Elastic search.

This is my query, please help me out

**Reply by reshmi-nair** [reply]
@Shakthieshwari You can either use user read api to fetch single user data at a time (which internally uses DB) or user search api to fetch multiple user data (which internally uses ES).  If it is for reports, you can use the redis user cache to fetch the details
